### DISCUSSION TITLE
Format table based on caption

### DISCUSSION THREAD

**Discussion by lettuceDestroyer** [typst/typst#4004] [Q&A] [answered]
I was wondering if it is possible to format a table based on a caption.

What I tried is something like this:
```typst
  #show table.cell.where(y: 0).and(figure.caption.content.contains("Tabletype 1")) : set text(weight: "bold")
```

But this throws the error ``function `caption` does not contain field `content` ``

Additionally I tried to make the table header bold with the following code which didn't work as excepted. 
```typst
  #show table.header : it => set text(weight: "bold")
```

**Comment by Mc-Zen** [comment]
Hi, this is basically possible by adding a condition to the body of the show rule:
```typ
#show table.cell.where(y: 0): it => {
  if it.body.has("text") and it.body.text.contains("TableType 1") {
    set text(weight: "bold")
    it
  } else {
    it
  }
}
```

Note that you should check whether the cells body actually has the field `text`. Otherwise, you'll get an error, when some other cell isn't made of just one text node and has some math as well for example. 

But I guess using a show-rule to perform this action will only be worth if you'd be typing this very often. Another sensible approach might be just to declare a variable:
```typ
let typetype1 = text(weight: "bold")[TableType 1]
```
and use it in the table.

**Reply by lettuceDestroyer** [reply]
This was really useful, thanks. While trying to modify the table a bit more I realised that all the fields are immutable. Do you think I could open an issue asking to make them mutable so that in the future we could use something like this?

```typst
#show figure.where(kind: table) : it => {
  if it.caption.body.text.contains("TableType 1") {
      // following line throws error: "cannot mutate fields on content"
      // it.body.align = left;
      
      // following three lines throw error: "cannot mutate fields on content"
      // for cell in it.body.children.at(0).children {
        // cell.body = strong(cell.body.text)
      //}
  } else {
    it
  }
}
```

**Reply by Mc-Zen** [reply]
Yes, they are immutable and as far as I can see this will definitely stay this way. Immutability has many advantages and allows the compiler to do stuff that were impossible otherwise. 

Rather, you just have to change your viewpoint a bit in approaching this, I would say. It takes some getting used to (I sure can tell) but pays off and makes the code very clear usually. 

Before I answer: Do I understand correctly, that you want to align the caption to the left? Or the entire table?

**Reply by Mc-Zen** [reply]
I might add, that idea of `show` rules is not to _modify_ content directly (that is _in-place_), but rather to enhance it (for example wrapping it in `strong()`, or replace it (i.e., by using the individual fields like in `figure` and recombining them). Usually, you can still achieve the same, just in a different way :)

**Reply by lettuceDestroyer** [reply]
What I tried to do is:
- make the headers bold
- align the whole table to the left (including the caption)

Regarding immutability having many advantages, would you enlighten me and tell me what those are?

**Reply by Mc-Zen** [reply, marked as answer]
This can be achieved with the following code
```typ
#show figure.where(kind: table) : it => {
  if it.caption.body.text.contains("TableType 1") {
      show table.cell.where(y: 0): strong
      // set align(center) // this is what we need to omit 
      it.body
      v(it.gap, weak: true)
      it.caption
  } else {
    it
  }
}
```
1. First, we select upon the caption containing `"TableType 1"`. 
2. Now we employ a show rule that will make the table headers bold. 
3. Finally, we build up the figure by recreating almost the original figure setup. We just omit the alignment. 

Note that this code has one caveat: Now, `figure(placement: auto/top/bottom)` will not work anymore for this table. This is unfortunate but probably you don't need it anyway. I'm not sure there even is a solution for this right now.

**Reply by Mc-Zen** [reply]
> Regarding immutability having many advantages, would you enlighten me and tell me what those are?

I haven't worked on the internals of Typst, I just write Typst code, so I don't quite know the mechanics of it but here is a motivation:
You may have noticed that Typst functions are pure, i.e., they cannot modify anything outside their scope. This would be called "side-effects". With such side-effects, the incremental compiler wouldn't know what exactly might need to be updated but if every function is self-contained and only affects the world around by its return value, changes can be tracked. This is partially what makes Typst so fast. 

Immutable things are of course generally much easier to reason about which is good to prevent bugs and to enable further optimization.

**Reply by lettuceDestroyer** [reply]
Thank you so much, you're a genius. 😸

**Reply by Mc-Zen** [reply]
You're welcome!
